galileo/src/layer/vector_tile_layer/expressions.rs

```diff
@@ -0,0 +1,443 @@
+use serde::{Deserialize, Serialize};
+
+use crate::error::GalileoError;
+use crate::Color;
+
+/// Arguments for exponential interpolation
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct LinearArgs<T> {
+    step_values: Option<Vec<StepValue<T>>>,
+}
+
+/// Arguments for exponential interpolation
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct StepArgs<T> {
+    step_values: Option<Vec<StepValue<T>>>,
+}
+
+/// Generic arguments for interpolation
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
+pub struct InterpolationArgs<T> {
+    base: Option<i32>,
+    /// Step values is internally sorted
+    step_values: Option<Vec<StepValue<T>>>,
+}
+#[derive(Copy, Clone, Debug, Serialize, Deserialize, PartialEq)]
+pub struct StepValue<T> {
+    resolution: f64,
+    step_value: T,
+}
+
+/// Type used to define expressions for interpolation
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
+pub struct InterpolateExpression<T> {
+    interpolation_type: Interpolation,
+    interpolation_args: InterpolationArgs<T>,
+}
+/// Type used to define steps
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
+pub struct StepExpression<T> {
+    default_value: T,
+    /// Each stop value maps the resolution to the T type
+    /// If, the current resolution is greater than step resolution
+    /// the value T maps to the T value where the step resolution is
+    /// less than that of current resolution.
+    step_values: Option<Vec<StepValue<T>>>,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
+pub struct ResolutionValueRange<T> {
+    max_resolution: f64,
+    min_resolution: f64,
+    start_value: Option<T>,
+    end_value: Option<T>,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
+pub struct ValueStep<T> {
+    resolution: f64,
+    value: T,
+}
+
+#[derive(Copy, Clone, Debug, Serialize, Deserialize, PartialEq)]
+pub enum Interpolation {
+    /// Linear interpolation type with base 1
+    Linear,
+    /// Exponential interpolation with variable base
+    Exponential,
+    /// Cubic Bezier interpolation
+    Cubic,
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
+pub enum StyleValue<T: Ord> {
+    /// Simple values like using Color::BLACK
+    Simple(T),
+    /// Interpolate expression
+    Interpolate(InterpolateExpression<T>),
+    /// Step expression
+    Steps(StepExpression<T>),
+}
+
+impl Color {
+    /// Evaluates value from simple color
+    pub fn get_value(&self) -> Result<Color, GalileoError> {
+        Ok(*self)
+    }
+}
+
+impl InterpolateExpression<Color> {
```

Comment:
You will want to refactor these, since we need to support interpolation not only for colors, but also for numeric values. And duplicating the mathematical functions for different types of arguments is never a good idea. If you think about the color as about just 4 different numbers, you'll see how it can be simplified. Also, please check if there are existing crates that provide implementation for interpolation functions.